### DISCUSSION TITLE
Lesson 7: forge test not calling external functions

### DISCUSSION THREAD

**Discussion by ishaan0x** [Cyfrin/foundry-full-course-cu#74] [Q&A] [answered]
I tried testing with `forge test` for the first time on Lesson 7. I copied FundMe.sol and PriceConverter.sol directly from the github.

However, when I did `forge test`, I got this error:

```
[FAIL. Reason: EvmError: Revert] testMinimumDollarIsFive() (gas: 5121)
Traces:
  [5121] FundMeTest::testMinimumDollarIsFive() 
    ├─ [0] 0x0000000000000000000000000000000000000000::MINIMUM_USD() [staticcall]
    │   └─ ← ()
    └─ ← "EvmError: Revert"

Test result: FAILED. 1 passed; 1 failed; finished in 3.41ms

Failing tests:
Encountered 1 failing test in test/FundMe.t.sol:FundMeTest
[FAIL. Reason: EvmError: Revert] testMinimumDollarIsFive() (gas: 5121)
```

My FundMe.t.sol:

```
contract FundMeTest is Test {
    FundMe fundMe;

    function setUpExternal() external {
        fundMe = new FundMe();
    }

    function testMinimumDollarIsFive() public {
        console.log(fundMe.MINIMUM_USD());
        assertEq(fundMe.MINIMUM_USD(), 5e18);
    }
}
```

I was able to fix it by changing FundMe.t.sol to this:
```
contract FundMeTest is Test {
    FundMe fundMe = new FundMe();

    function setUpExternal() external {
        fundMe = new FundMe();
    }

    function testMinimumDollarIsFive() public {
        console.log(fundMe.MINIMUM_USD());
        assertEq(fundMe.MINIMUM_USD(), 5e18);
    }
}
```

But [in the video](https://youtu.be/sas02qSFZ74?t=1200), the original FundMe.t.sol works. What am I doing wrong?

**Comment by ishaan0x** [comment]
also thank you Patrick for making these videos you're an absolute goat

**Comment by alymurtazamemon** [comment, marked as answer]
@ishaanh You are using a different setup function name due to that the foundry executor to able to identify it.
```solidity
function setUpExternal() external {
    fundMe = new FundMe();
}
```
it can only identify the setup function with the name `setUp`. And due to that your fund me contract's instance was empty.

**Reply by cromewar** [reply]
Dos this make sense for you @ishaanh?

**Reply by ishaan0x** [reply]
Yes this works! Thanks so much @alymurtazamemon 
Do you know where we can find a list of these special functions (where the function only works if it's a specific name)? I know of `constructor` `fallback` `receive` and now `setUp`.

**Reply by ghost** [reply]
There is a quick explanation about the `setUp` function in the [Writing Tests](https://book.getfoundry.sh/forge/writing-tests) section of the docs.

The `run` function is another function constantly used by Patrick in the videos. The section [Solidity Scripting](https://book.getfoundry.sh/tutorials/solidity-scripting#writing-the-script) has a good example and talks a little bit about this function.

But remember: 
- `setUp` and `run` are "special functions" in Foundry, the framework. 
- `constructor`, `fallback` and `receive` are special functions in Solidity, the language.

**Reply by alymurtazamemon** [reply]
@ishaanh Documentation!
